src/Control/Carrier/Writer/Strict.hs

```diff
@@ -47,25 +49,20 @@ execWriter = fmap fst . runWriter
 
 
 -- | A space-efficient carrier for 'Writer' effects, implemented atop "Control.Carrier.State.Strict".
-newtype WriterC w m a = WriterC { runWriterC :: StateC w m a }
+newtype WriterC w m a = WriterC (StateC w m a)
   deriving (Alternative, Applicative, Functor, Monad, Fail.MonadFail, MonadFix, MonadIO, MonadPlus, MonadTrans)
 
 instance (Monoid w, Carrier sig m, Effect sig) => Carrier (Writer w :+: sig) (WriterC w m) where
-  eff (L (Tell w     k)) = WriterC $ do
-    modify (`mappend` w)
-    runWriterC k
-  eff (L (Listen   m k)) = WriterC $ do
-    w <- get
-    put (mempty :: w)
-    a <- runWriterC m
-    w' <- get
-    modify (mappend (w :: w))
-    runWriterC (k w' a)
-  eff (L (Censor f m k)) = WriterC $ do
-    w <- get
-    put (mempty :: w)
-    a <- runWriterC m
-    modify (mappend w . f)
-    runWriterC (k a)
+  eff (L (Tell w     k)) = WriterC (modify (`mappend` w)) >> k
+  eff (L (Listen   m k)) = WriterC (StateC (\ w -> do
+    (w', a) <- runWriter m
+    let w'' = mappend w w'
+    w'' `seq` pure (w'', (w', a))))
+    >>= uncurry k
+  eff (L (Censor f m k)) = WriterC (StateC (\ w -> do
+    (w', a) <- runWriter m
+    let w'' = mappend w (f w')
+    w'' `seq` pure (w'', a)))
+    >>= k
```

Comment:
This was largely rewritten to enable us to bind the continuations in `WriterC` instead of in `m`, but as a bonus, it’s quite a bit tidier due to directly manipulating the underlying `StateC`.